usr.bin/ldd/ldd.1
.\"
.Dd March 21, 2023
.Dt LDD 1
.Os
.Sh NAME
.Nm ldd
.Nd list dynamic object dependencies
.Sh SYNOPSIS
.Nm
.Op Fl a
.Op Fl f Ar format Op Fl f Ar format
.Ar program ...
.Sh DESCRIPTION
The
.Nm
utility displays all shared objects that are needed to run the given program or
to load the given shared object.
Contrary to
.Xr nm 1 ,
the list includes
.Dq indirect
dependencies that are the result of needed shared objects which themselves
depend on yet other shared objects.
.Pp
Zero, one or two
.Fl f
options may be given.
The argument is a format string passed to
.Xr rtld 1
and allows customization of
.Nm Ns 's
output.
If one is given, it sets
.Ev LD_TRACE_LOADED_OBJECTS_FMT1 .
If two are given, they set
.Ev LD_TRACE_LOADED_OBJECTS_FMT1
and
.Ev LD_TRACE_LOADED_OBJECTS_FMT2 ,
respectively.
See
.Xr rtld 1
for details, including a list of recognized conversion characters.
.Pp
The
.Fl a
option displays the list of all objects that are needed by each loaded
object.
.Sh IMPLEMENTATION NOTES
.Nm
lists the dependencies of an executable by setting
.Xr rtld 1
environment variables and running the executable in a child process.
If the executable is corrupt or invalid,
.Nm
may therefore fail without providing any diagnostic error messages.
.Sh EXAMPLES
The following is an example of a shell pipeline which uses the
.Fl f
option.
It will print a report of all ELF binaries in the current directory,
which link against libc.so.6:
.Dl "find . -type f | xargs -n1 file -F ' ' | grep 'ELF.*dynamically' | cut -f1 -d' ' | xargs ldd -f '%A %o\en' | grep libc.so.6"
.Sh SEE ALSO
.Xr ld 1 ,
.Xr nm 1 ,
.Xr readelf 1 ,
.Xr rtld 1
.Sh HISTORY
A
.Nm
utility first appeared in SunOS 4.0, it appeared in its current form in
.Fx 1.1 .


NEWS
Dec 07, 2022: libarchive 3.6.2 released

Apr 08, 2022: libarchive 3.6.1 released

Feb 09, 2022: libarchive 3.6.0 released

Feb 08, 2022: libarchive 3.5.3 released

Aug 22, 2021: libarchive 3.5.2 released

Dec 26, 2020: libarchive 3.5.1 released

Dec 01, 2020: libarchive 3.5.0 released

Oct 14, 2020: Support for system extended attributes

May 20, 2020: libarchive 3.4.3 released

Apr 30, 2020: Support for pzstd compressed files

Apr 16, 2020: Support for RHT.security.selinux tar extended attribute

Feb 11, 2020: libarchive 3.4.2 released

Jan 23, 2020: Important fixes for writing XAR archives

Jan 20, 2020: New tar option: --safe-writes (atomical file extraction)

Jan 03, 2020: Support mbed TLS (PolarSSL) as optional crypto provider

Dec 30, 2019: libarchive 3.4.1 released

Dec 11, 2019: New pax write option "xattrhdr"

Nov 17, 2019: Unicode filename support for reading lha/lzh archives

Jun 11, 2019: libarchive 3.4.0 released

May 18, 2019: Fixes for reading Android APK and JAR archives

Apr 16, 2019: Support for non-recursive list and extract

Apr 14, 2019: New tar option: --exclude-vcs

Mar 27, 2019: Support for file and directory symlinks on Windows

Mar 12, 2019: Important fixes for storing file attributes and flags

Jan 20, 2019: Support for xz, lzma, ppmd8 and bzip2 decompression in ZIP files

Oct 06, 2018: RAR 5.0 reader

Sep 03, 2018: libarchive 3.3.3 released

Jul 19, 2018: Avoid super-linear slowdown on malformed mtree files

Jan 27, 2018: Many fixes for building with Visual Studio

Oct 19, 2017: NO_OVERWRITE doesn't change existing directory attributes

Aug 12, 2017: New support for Zstandard read and write filters

Jul 09, 2017: libarchive 3.3.2 released

Mar 16, 2017: NFSv4 ACL support for Linux (librichacl)

Feb 26, 2017: libarchive 3.3.1 released
    Security & Feature release

Feb 19, 2017: libarchive 3.3.0 released
    Security & Feature release

Jan 29, 2017: Limited NFSv4 ACL support for Mac OS (Darwin)

Jan 10, 2017: POSIX.1e and NFSv4 ACL support for Solaris and derivates

Dec 27, 2016: NFSv4 ACL read and write support for pax
    Deprecated functions: archive_entry_acl_text(), archive_entry_acl_text_w()

Nov, 2016: libarchive is now being tested by the OSS-Fuzz project

Oct 26, 2016: Remove liblzmadec support

Oct 23, 2016: libarchive 3.2.2 released
    Security release

Jun 20, 2016: libarchive 3.2.1 released
    This fixes a handful of security and other critical issues with 3.2.0

May 01, 2016: libarchive 3.2.0 released

Apr 09, 2016: libarchive 3.1.901a released
    Another test release in preparation for 3.2.0

Feb 13, 2016: libarchive 3.1.900a released
    This is a test release in preparation for 3.2.0

Oct 21, 2015: Preliminary port to OSF

Apr 11, 2015: libarchive's issue tracker is now hosted at GitHub.
    https://github.com/libarchive/libarchive/issues

Early 2015:  Many fixes to crash and overflow bugs thanks to Hanno Boeck

Oct 13, 2014: Zip encryption and decryption support

Aug 13, 2014: Add support for lz4 compression.

Jun 10, 2014: Add warc format support

May  3, 2014: Add experimental Zip streaming extension

Apr  6, 2014: Add bsdcat command-line tool

Jan 12, 2014: Add Zip64 support

Dec  1, 2013: Rewrite Zip write logic

Jul  1, 2013: Add ability to detect encrypted entries for many formats
     (This does not add the ability to *decrypt* those entries, however)

Feb 23, 2013: "raw" write support added

Feb 09, 2013: libarchive 3.1.2 released

Jan 28, 2013: libarchive's new website moved to http://www.libarchive.org.

Jan 13, 2013: libarchive 3.1.1 released

Jan 13, 2013: libarchive 3.1.0 released

Dec 07, 2012: Implement functions to manually set the format and filters used.
    
Nov 11, 2012: Add support for __MACOSX directory in Zip archives, which resource
    forks are stored in.

Oct 20, 2012: Add support for writing v7 tar format.

Oct 09, 2012: Add support for grzip compression.

Oct 07, 2012: Introduce b64encode filter.    
Oct 07, 2012: Introduce uuencode filter.

Oct 06, 2012: Add support for lzop.
    
Sep 27, 2012: Implement function used to seek within data blocks.
    (Currently only supported for uncompressed RAR archives).
    
Apr 22, 2012: Add basic archive read and write filter support for lrzip.

Mar 27, 2012: libarchive 3.0.4 released

Feb 05, 2012: libarchive development now hosted at GitHub.
    http://libarchive.github.com/
Feb 05, 2012: libarchive's issue tracker remains at Google Code.
    http://code.google.com/p/libarchive/issues/list
Feb 05, 2012: libarchive's mailing lists remain at Google Groups.

Dec 24, 2011: libarchive 3.0.2 released
Dec 23, 2011: Various fixes merged from FreeBSD
Dec 23, 2011: Symlink support in Zip reader and writer
Dec 23, 2011: Robustness fixes to 7Zip reader

Nov 27, 2011: libarchive 3.0.1b released

Nov 26, 2011: 7Zip reader
Nov 26, 2011: Small fixes to ISO and Zip to improve robustness with corrupted input
Nov 24, 2011: Improve streaming Zip reader's support for uncompressed entries
Nov 20, 2011: New seeking Zip reader supports SFX Zip archives
Nov 20, 2011: Build fixes on Windows

Nov 13, 2011: libarchive 3.0.0a released

Nov 06, 2011: Update shared-library version calculations for libarchive 3.x
Sep 04, 2011: Fix tar -s; follow GNU tar for controlling hardlink/symlink substitutions
Aug 18, 2011: Fix reading ISO images built by NetBSD's mkisofs
Aug 15, 2011: Old archive_read_support_compression_XXX functions are deprecated and
   will disappear in libarchive 4.0.
Jun 26, 2011: RAR reader
Jun 16, 2011: Add tar:compat-2x option to emulate broken libarchive 2.x
   handling of pax UTF-8 headers
Apr 25, 2011: Refactor read_open() into a collection of single-item setters;
   support the old interfaces as wrappers
Apr 12, 2011: Split disk writer into separate POSIX and Windows implementations
Apr 10, 2011: Improvements to character translations on Windows.
Mar 30, 2011: More work to return errors instead of calling abort()

usr.bin/ldd/ldd.c
/*-
 * SPDX-License-Identifier: BSD-4-Clause
 *
 * Copyright (c) 1993 Paul Kranenburg
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *      This product includes software developed by Paul Kranenburg.
 * 4. The name of the author may not be used to endorse or promote products
 *    derived from this software without specific prior written permission
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#include <sys/param.h>
#include <sys/wait.h>

#include <machine/elf.h>

#include <arpa/inet.h>

#include <dlfcn.h>
#include <err.h>
#include <errno.h>
#include <fcntl.h>
#include <gelf.h>
#include <libelf.h>
#include <rtld_paths.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

/*
 * 32-bit ELF data structures can only be used if the system header[s] declare
 * them.  There is no official macro for determining whether they are declared,
 * so check for the existence of one of the 32-macros defined in elf(5).
 */
#ifdef ELF32_R_TYPE
#define	ELF32_SUPPORTED
#endif

#define	LDD_SETENV(name, value, overwrite) do {		\
	setenv("LD_" name, value, overwrite);		\
	setenv("LD_32_" name, value, overwrite);	\
} while (0)

#define	LDD_UNSETENV(name) do {		\
	unsetenv("LD_" name);		\
	unsetenv("LD_32_" name);	\
} while (0)

static int	is_executable(const char *fname, int fd, int *is_shlib,
		    int *type);
static void	usage(void);

#define	TYPE_UNKNOWN	0
#define	TYPE_ELF	1	/* Architecture default */
#if __ELF_WORD_SIZE > 32 && defined(ELF32_SUPPORTED)
#define	TYPE_ELF32	2	/* Explicit 32 bits on architectures >32 bits */

#define	_PATH_LDD32	"/usr/bin/ldd32"

static int
execldd32(char *file, char *fmt1, char *fmt2, int aflag)
{
	char *argv[9];
	int i, rval, status;

	LDD_UNSETENV("TRACE_LOADED_OBJECTS");
	rval = 0;
	i = 0;
	argv[i++] = strdup(_PATH_LDD32);
	if (aflag)
		argv[i++] = strdup("-a");
	if (fmt1 != NULL) {
		argv[i++] = strdup("-f");
		argv[i++] = strdup(fmt1);
	}
	if (fmt2 != NULL) {
		argv[i++] = strdup("-f");
		argv[i++] = strdup(fmt2);
	}
	argv[i++] = strdup(file);
	argv[i++] = NULL;

	switch (fork()) {
	case -1:
		err(1, "fork");
		break;
	case 0:
		execv(_PATH_LDD32, argv);
		warn("%s", _PATH_LDD32);
		_exit(127);
		break;
	default:
		if (wait(&status) < 0)
			rval = 1;
		else if (WIFSIGNALED(status))
			rval = 1;
		else if (WIFEXITED(status) && WEXITSTATUS(status) != 0)
			rval = 1;
		break;
	}
	while (i--)
		free(argv[i]);
	LDD_SETENV("TRACE_LOADED_OBJECTS", "yes", 1);
	return (rval);
}
#endif

int
main(int argc, char *argv[])
{
	char *fmt1, *fmt2;
	const char *rtld;
	int aflag, c, fd, rval, status, is_shlib, rv, type;

	aflag = 0;
	fmt1 = fmt2 = NULL;

	while ((c = getopt(argc, argv, "af:")) != -1) {
		switch (c) {
		case 'a':
			aflag++;
			break;
		case 'f':
			if (fmt1 != NULL) {
				if (fmt2 != NULL)
					errx(1, "too many formats");
				fmt2 = optarg;
			} else
				fmt1 = optarg;
			break;
		default:
			usage();
			/* NOTREACHED */
		}
	}
	argc -= optind;
	argv += optind;

	if (argc <= 0) {
		usage();
		/* NOTREACHED */
	}

	rval = 0;
	for (; argc > 0; argc--, argv++) {
		if ((fd = open(*argv, O_RDONLY | O_VERIFY, 0)) < 0) {
			warn("%s", *argv);
			rval |= 1;
			continue;
		}
		rv = is_executable(*argv, fd, &is_shlib, &type);
		close(fd);
		if (rv == 0) {
			rval |= 1;
			continue;
		}

		switch (type) {
		case TYPE_ELF:
			break;
#if __ELF_WORD_SIZE > 32 && defined(ELF32_SUPPORTED)
		case TYPE_ELF32:
			rval |= execldd32(*argv, fmt1, fmt2, aflag);
			continue;
#endif
		case TYPE_UNKNOWN:
		default:
			/*
			 * This shouldn't happen unless is_executable()
			 * is broken.
			 */
			errx(EDOOFUS, "unknown executable type");
		}

		/* ld.so magic */
		LDD_SETENV("TRACE_LOADED_OBJECTS", "yes", 1);
		if (fmt1 != NULL)
			LDD_SETENV("TRACE_LOADED_OBJECTS_FMT1", fmt1, 1);
		if (fmt2 != NULL)
			LDD_SETENV("TRACE_LOADED_OBJECTS_FMT2", fmt2, 1);

		LDD_SETENV("TRACE_LOADED_OBJECTS_PROGNAME", *argv, 1);
		if (aflag)
			LDD_SETENV("TRACE_LOADED_OBJECTS_ALL", "1", 1);
		else if (fmt1 == NULL && fmt2 == NULL)
			/* Default formats */
			printf("%s:\n", *argv);
		fflush(stdout);

		switch (fork()) {
		case -1:
			err(1, "fork");
			break;
		default:
			if (wait(&status) < 0) {
				warn("wait");
				rval |= 1;
			} else if (WIFSIGNALED(status)) {
				fprintf(stderr, "%s: signal %d\n", *argv,
				    WTERMSIG(status));
				rval |= 1;
			} else if (WIFEXITED(status) &&
			    WEXITSTATUS(status) != 0) {
				fprintf(stderr, "%s: exit status %d\n", *argv,
				    WEXITSTATUS(status));
				rval |= 1;
			}
			break;
		case 0:
			rtld = _PATH_RTLD;
#if __ELF_WORD_SIZE > 32 && defined(ELF32_SUPPORTED)
			if (type == TYPE_ELF32)
				rtld = _COMPAT32_PATH_RTLD;
#endif
			if (is_shlib == 0) {
				execl(rtld, rtld, "--",
				    *argv, (char *)NULL);
				warn("%s", *argv);
			} else if (fmt1 == NULL && fmt2 == NULL && !aflag) {
				dlopen(*argv, RTLD_TRACE);
				warnx("%s: %s", *argv, dlerror());
			} else {
				execl(rtld, rtld, "-d", "--",
				    *argv, (char *)NULL);
			}
			_exit(1);
		}
	}

	return (rval);
}

static void
usage(void)
{

	fprintf(stderr,
	    "usage: ldd [-a] [-f format [-f format]] program ...\n");
	exit(1);
}

static bool
has_freebsd_abi_tag(const char *fname, Elf *elf, GElf_Ehdr *ehdr, off_t offset,
    size_t len)
{
	Elf_Data dst, src;
	const Elf_Note *note;
	char *buf;
	const char *name;
	void *copy;
	size_t namesz, descsz;
	bool has_abi_tag;

	buf = elf_rawfile(elf, NULL);
	if (buf == NULL) {
		warnx("%s: %s", fname, elf_errmsg(0));
		return (false);
	}

	memset(&src, 0, sizeof(src));
	src.d_buf = buf + offset;
	src.d_size = len;
	src.d_type = ELF_T_NOTE;
	src.d_version = EV_CURRENT;

	memset(&dst, 0, sizeof(dst));
	dst.d_buf = copy = malloc(len);
	dst.d_size = len;
	dst.d_type = ELF_T_NOTE;
	dst.d_version = EV_CURRENT;

	if (gelf_xlatetom(elf, &dst, &src, ehdr->e_ident[EI_DATA]) == NULL) {
		warnx("%s: failed to parse notes: %s", fname, elf_errmsg(0));
		free(copy);
		return (false);
	}

	buf = copy;
	has_abi_tag = false;
	for (;;) {
		if (len < sizeof(*note))
			break;

		note = (const void *)buf;
		buf += sizeof(*note);
		len -= sizeof(*note);

		namesz = roundup2(note->n_namesz, sizeof(uint32_t));
		descsz = roundup2(note->n_descsz, sizeof(uint32_t));
		if (len < namesz + descsz)
			break;

		name = buf;
		if (note->n_namesz == sizeof(ELF_NOTE_FREEBSD) &&
		    strncmp(name, ELF_NOTE_FREEBSD, note->n_namesz) == 0 &&
		    note->n_type == NT_FREEBSD_ABI_TAG &&
		    note->n_descsz == sizeof(uint32_t)) {
			has_abi_tag = true;
			break;
		}

		buf += namesz + descsz;
		len -= namesz + descsz;
	}

	free(copy);
	return (has_abi_tag);
}

static bool
is_pie(const char *fname, Elf *elf, GElf_Ehdr *ehdr, off_t offset, size_t len)
{
	Elf_Data dst, src;
	char *buf;
	void *copy;
	const GElf_Dyn *dyn;
	size_t dynsize;
	u_int count, i;
	bool pie;

	buf = elf_rawfile(elf, NULL);
	if (buf == NULL) {
		warnx("%s: %s", fname, elf_errmsg(0));
		return (false);
	}

	dynsize = gelf_fsize(elf, ELF_T_DYN, 1, EV_CURRENT);
	if (dynsize == 0) {
		warnx("%s: %s", fname, elf_errmsg(0));
		return (false);
	}
	count = len / dynsize;

	memset(&src, 0, sizeof(src));
	src.d_buf = buf + offset;
	src.d_size = len;
	src.d_type = ELF_T_DYN;
	src.d_version = EV_CURRENT;

	memset(&dst, 0, sizeof(dst));
	dst.d_buf = copy = malloc(count * sizeof(*dyn));
	dst.d_size = count * sizeof(*dyn);
	dst.d_type = ELF_T_DYN;
	dst.d_version = EV_CURRENT;

	if (gelf_xlatetom(elf, &dst, &src, ehdr->e_ident[EI_DATA]) == NULL) {
		warnx("%s: failed to parse .dynamic: %s", fname, elf_errmsg(0));
		free(copy);
		return (false);
	}

	dyn = copy;
	pie = false;
	for (i = 0; i < count; i++) {
		if (dyn[i].d_tag != DT_FLAGS_1)
			continue;

		pie = (dyn[i].d_un.d_val & DF_1_PIE) != 0;
		break;
	}

	free(copy);
	return (pie);
}

static int
is_executable(const char *fname, int fd, int *is_shlib, int *type)
{
	Elf *elf;
	GElf_Ehdr ehdr;
	GElf_Phdr phdr;
	bool dynamic, freebsd, pie;
	int i;

	*is_shlib = 0;
	*type = TYPE_UNKNOWN;
	dynamic = false;
	freebsd = false;
	pie = false;

	if (elf_version(EV_CURRENT) == EV_NONE) {
		warnx("unsupported libelf");
		return (0);
	}
	elf = elf_begin(fd, ELF_C_READ, NULL);
	if (elf == NULL) {
		warnx("%s: %s", fname, elf_errmsg(0));
		return (0);
	}
	if (elf_kind(elf) != ELF_K_ELF) {
		elf_end(elf);
		warnx("%s: not a dynamic ELF executable", fname);
		return (0);
	}
	if (gelf_getehdr(elf, &ehdr) == NULL) {
		warnx("%s: %s", fname, elf_errmsg(0));
		elf_end(elf);
		return (0);
	}

	*type = TYPE_ELF;
#if __ELF_WORD_SIZE > 32 && defined(ELF32_SUPPORTED)
	if (gelf_getclass(elf) == ELFCLASS32) {
		*type = TYPE_ELF32;
	}
#endif

	freebsd = ehdr.e_ident[EI_OSABI] == ELFOSABI_FREEBSD;
	for (i = 0; i < ehdr.e_phnum; i++) {
		if (gelf_getphdr(elf, i, &phdr) == NULL) {
			warnx("%s: %s", fname, elf_errmsg(0));
			elf_end(elf);
			return (0);
		}
		switch (phdr.p_type) {
		case PT_NOTE:
			if (ehdr.e_ident[EI_OSABI] == ELFOSABI_NONE && !freebsd)
				freebsd = has_freebsd_abi_tag(fname, elf, &ehdr,
				    phdr.p_offset, phdr.p_filesz);
			break;
		case PT_DYNAMIC:
			dynamic = true;
			if (ehdr.e_type == ET_DYN)
				pie = is_pie(fname, elf, &ehdr, phdr.p_offset,
				    phdr.p_filesz);
			break;
		}
	}

	if (!dynamic) {
		elf_end(elf);
		warnx("%s: not a dynamic ELF executable", fname);
		return (0);
	}

	if (ehdr.e_type == ET_DYN && !pie) {
		*is_shlib = 1;

		if (!freebsd) {
			elf_end(elf);
			warnx("%s: not a FreeBSD ELF shared object", fname);
			return (0);
		}
	}

	elf_end(elf);
	return (1);
}
